iAPI: Refactor types and add a "Core Concepts - Using TypeScript" guide (#64577)
WordPress/gutenberg: docs/reference-guides/interactivity-api/core-concepts/using-typescript.md

    ```ts
    const { state, actions } = store( 'myCounterPlugin', {
    	state: {
    		counter: 0,
    	},
    	actions: {
    		*delayedReturn(): Generator< uknown, number, uknown > {
    			yield new Promise( ( r ) => setTimeout( r, 1000 ) );
    			return state.counter; // Now this is correctly inferred.
    		},
    	},
    } );
    ```

    That's it! Remember that the return type of a Generator is the second generic argument: `Generator< unknown, ReturnType, unknown >`.

2.  When yielding inside a generator, TypeScript does not resolve the types of the promises directly because it doesn't know what the controller will return after processing the promise.

    ```ts
    const getAsyncNumber = async () => {
    	return 1;
    };

    store( 'myCounterPlugin', {
    	actions: {
    		*getSomeNumber() {
    			const n = yield getAsyncNumber(); // TypeScript doesn't resolve this type.
    		},
    	},
    } );
    ```

    To fix this, you can combine the use of `yield*` with the `typed()` helper, which converts the promise type into an iterator that can be correctly inferred by TypeScript when used in conjunction with `yield*`.

>     _**Note**: The `typed` function will be included in WordPress 6.7._

Reviewer: It's fine to land it earlier as:

```suggestion
    _**Note**: The `typed` function introduced in WordPress 6.7._
```

This is a common practice to avoid unnecessary updates after major releases.

    To fix this, you can combine the use of `yield*` with the `typed()` helper, which converts the promise type into an iterator that can be correctly inferred by TypeScript when used in conjunction with `yield*`.

>     _**Note**: The `typed` function will be included in WordPress 6.7._

Author: When WordPress 6.7 is released, we should update this note to something like this: `The typed function was included in WordPress 6.7.`
